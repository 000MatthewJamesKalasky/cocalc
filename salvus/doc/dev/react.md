# React.js style conventions for SMC



* ref should be the first prop listed.

  ref names the component, e.g. `<Input ref = 'foo' type = 'text' />` you can then access the Input with something like `@refs.foo.getValue()`

* non-react functions should follow `python_naming` conventions.

* never call an action and immediately access those props from the store. The dispatcher may not be updated immediately!

* recommend keeping functions shorter than 1 page.
  1 page == 50 lines?



> hsy: more generally, does anyone know if there is
> a tool to reformat the source code? e.g. for other
> python projects, I'm running autopep8 across all .py files

> to maybe answer my own question:
> https://github.com/emorikawa/coffeelint-cjsx/

> from their main website, one can pick some style rules
> http://www.coffeelint.org/