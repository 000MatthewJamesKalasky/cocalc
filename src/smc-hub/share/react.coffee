###
Load react support for rendering.
###

# Code for rendering react components to html.
ReactDOMServer = require('react-dom/server')

require('./jsdom-support')

{mathjax} = require('./mathjax-support')

# Uncomment for cc-in-cc dev benchmarking purposes.  This variable is already set
# by the Docker container when running in kubernetes.
## process.NODE_ENV="production"


STREAMING = false
# Ned to implement more targetted mathjax before enabling streaming

if STREAMING
    # We use streaming rendering -- see https://hackernoon.com/whats-new-with-server-side-rendering-in-react-16-9b0d78585d67
    exports.react = (res, component, extra) ->
        t0 = new Date()
        res.type('html')
        stream = ReactDOMServer.renderToStaticNodeStream(component)
        stream.once 'end', ->
            console.log("react: time to render and send: #{new Date() - t0}ms", extra)
        stream.pipe(res)

else

    exports.react = (res, component, extra) ->
        t0 = new Date()
        html = '<!DOCTYPE html>' + ReactDOMServer.renderToStaticMarkup(component)
        if html.indexOf('cocalc-share-mathjax') != -1
            mathjax html, (err, html) ->
                console.log("react: time to render and send: #{new Date() - t0}ms", extra)
                res.send(html)
        else
            res.send(html)
