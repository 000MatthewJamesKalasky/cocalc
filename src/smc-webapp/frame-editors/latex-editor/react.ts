export const {
  React,
  ReactDOM,
  rclass,
  rtypes,
  Fragment
} = require("smc-webapp/smc-react");

import * as ReactOrig from "react";

export const Component = ReactOrig.Component;

export type Rendered = ReactOrig.ReactElement<any> | undefined;
