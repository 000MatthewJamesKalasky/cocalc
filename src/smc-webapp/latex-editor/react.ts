export const { React, ReactDOM, rclass, rtypes, Fragment } = require("../smc-react");

import * as ReactOrig from "react";

export const Component = ReactOrig.Component;