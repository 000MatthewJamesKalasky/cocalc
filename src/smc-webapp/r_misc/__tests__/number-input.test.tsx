import * as React from "react";
import { shallow } from "enzyme";
import { NumberInput } from "../number-input";

test("smoke test", () => {
  const starting_num = 0;
  const mock_change = jest.fn();
  const min = 10;
  const max = 20;

  const rendered = shallow(
    <NumberInput
      number={starting_num}
      min={min}
      max={max}
      on_change={mock_change}
    />
  );
  expect(rendered).toMatchSnapshot();
});
