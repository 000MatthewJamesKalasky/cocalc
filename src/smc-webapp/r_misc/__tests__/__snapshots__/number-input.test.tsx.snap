// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke test 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    xs={6}
  >
    <form
      onSubmit={[Function]}
    >
      <FormGroup
        bsClass="form-group"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          type="text"
          value={0}
        />
      </FormGroup>
    </form>
  </Col>
  <Col
    bsClass="col"
    className="lighten"
    componentClass="div"
    xs={6}
  />
</Row>
`;
